File: wintouch/FbCategoryMessage.cpp
#include "FbCategoryMessage.h"

#include <msclr\marshal_cppstd.h>

#include "MainWindow.h"

FbCategoryMessage::FbCategoryMessage(const TheUI::FbEventData& msg)
	: FbMessage(TableName(), msg)
	, name()
{ }

int FbCategoryMessage::Respond()
{
	int ret = -1;

	if (!isParsed && Parse() != 0)
	{
		ParseFailedMessage();
	}
	else if (OperationType() == TheUI::Fb_Operations::fb_add || OperationType() == TheUI::Fb_Operations::fb_edit)
	{
		System::String^ idCli = msclr::interop::marshal_as<System::String^>(id);
		System::String^ nameCli = msclr::interop::marshal_as<System::String^>(name);
		TheUI::CategoryPropEntryFb item(nameCli);
		ManagedCode::ManagedGlobals::w->AddProperties(idCli, %item);
		ret = 0;
	}
	else if (OperationType() == TheUI::Fb_Operations::fb_delete)
	{
		System::String^ idCli = msclr::interop::marshal_as<System::String^>(id);
		ManagedCode::ManagedGlobals::w->RemoveCategoryProperties(idCli);
		ret = 0;
	}
	else UnsupportedOperationMessage();

	return ret;
}

int FbCategoryMessage::Parse()
{
	if (OperationType() == TheUI::Fb_Operations::fb_add || OperationType() == TheUI::Fb_Operations::fb_edit)
	{
		if (!isParsed && ParseId(NAME)== 0) { name = Data(); }
	}
	else if (OperationType() == TheUI::Fb_Operations::fb_delete)
	{
		ParseId();
	}
	return isParsed ? 0 : -1;
}

const std::string FbCategoryMessage::NAME = "Name";

File: TheUI/FbMessages/FbCategoryMessage.cs
namespace TheUI
{
    class FbCategoryMessage : FbMessage, ICategoryPropEntryFb
    {
        public FbCategoryMessage() : base(new FbEventData()) { }
        public FbCategoryMessage(FbEventData msg) : base(msg)
        {
            Name = "";
            Parse();
        }

        public override string TABLE_NAME { get { return "categories"; } }

        protected override sealed int Parse()
        {
            if (OPERATION == Fb_Operations.fb_add || OPERATION == Fb_Operations.fb_edit)
            {
                if (!IsParsed && ParseId(NAME) == 0) { Name = DATA; }
            }
            else if (OPERATION == Fb_Operations.fb_delete)
            {
                ParseId();
            }
            return IsParsed ? 0 : -1;
        }

        public string Name { get; set; }

        private const string NAME = "Name";
    }


}
